# Sounds Data Paper Python

## Artists List

There are multiple lists of artists. One is in the `artistsIndex` worksheet. This should presumably be the master artist list since it contains indexes for all of them.

There are also artists in other worksheets: `Venue_Artist_Event` and `Ven_Ev_Address`. The first task is to double check artist names and ensure we have a complete list with unique indexes for all artists.

### Load and Inspect the Data

First, I'm going to load the excel file and access the worksheet.

Here, I'm going to import the pandas library, read the file, and inspect the names of the worksheets.

In [70]:
import pandas as pd

In [71]:
sounds_data = pd.ExcelFile('spreadsheets/product1_data_master_rev.xlsx')

In [72]:
sheets = sounds_data.sheet_names
sheets

['Ven_Ev_Ar_Address',
 'Venue_Artist_Event',
 'Layer 1 Venues',
 'artistsGenre',
 'genreIndex',
 'Sheet5',
 'artistsIndex',
 'Venues No Address',
 'venues_with_coordinates']

Now that I have the exact names of all the sheets, I will go ahead and load the artistsIndex sheet and the other sheets I am interested in. Then I will inspect the resulting pandas data frames to find the names of the columns containing artist information.

In [73]:
artists_index = pd.read_excel('spreadsheets/product1_data_master_rev.xlsx', sheet_name='artistsIndex')
venue_artist_event = pd.read_excel('spreadsheets/product1_data_master_rev.xlsx', sheet_name='Venue_Artist_Event')
ven_ev_ar_address = pd.read_excel('spreadsheets/product1_data_master_rev.xlsx', sheet_name='Ven_Ev_Ar_Address')

In [74]:
artists_index.head()

,Name,Index
0,Dierks Bentley,1
1,"""The Italian Cowboy""",2
2,Keith Urban,3
3,George Jones,4
4,Hilary Duff,5


In [75]:
venue_artist_event.head()

,Month,Day,Year,Venue,Event_Artists,Artist_Formula,Event_Formulas,Index
0,September,10,1857,St. Mary's Catholic Church,French Mountaineer Singers,French Mountaineer Singers,NaN,1
1,June,25,1869,Menger Hotel,General Mason Farewell Party: San Antonio Bras...,San Antonio Brass Band,General Mason Farewell Party,2
2,July,4,1874,San Pedro Park,San Antonio Brass Band,San Antonio Brass Band,NaN,3
3,October,9,1874,"San Pedro Springs, Turner Hall",Tenth Texas Saengerfest,Tenth Texas Saengerfest,NaN,4
4,October,10,1874,"Casino Hall, Turner Hall",Tenth Texas Saengerfest,Tenth Texas Saengerfest,NaN,5


In [76]:
ven_ev_ar_address.head()

,Index,Month,Day,Year,Venue,Event_Artists,Address,Zip,Longitude,Latitude,Artist,Event,City,State
0,2581,January,1,1970,San Antonio Convention Center Arena,"KBER Holiday Show: Willie Nelson, Charley Pri...",601 Hemisfair Plaza Way,78205.0,NaN,NaN,Willie Nelson,KBER Holiday Show,San Antonio,Texas
1,2581,January,1,1970,San Antonio Convention Center Arena,"KBER Holiday Show: Willie Nelson, Charley Pri...",601 Hemisfair Plaza Way,78205.0,NaN,NaN,Charley Pride,KBER Holiday Show,San Antonio,Texas
2,2581,January,1,1970,San Antonio Convention Center Arena,"KBER Holiday Show: Willie Nelson, Charley Pri...",601 Hemisfair Plaza Way,78205.0,NaN,NaN,Grandpa Jones,KBER Holiday Show,San Antonio,Texas
3,2581,January,1,1970,San Antonio Convention Center Arena,"KBER Holiday Show: Willie Nelson, Charley Pri...",601 Hemisfair Plaza Way,78205.0,NaN,NaN,David Houston,KBER Holiday Show,San Antonio,Texas
4,2582,January,16,1970,San Antonio Municipal Auditorium,Vikki Carr,100 Auditorium Cir.,78205.0,NaN,NaN,Vikki Carr,NaN,San Antonio,Texas


### Get the Set of Unique Artists

Now, I'm going to extract the list of unique artists listed in the artistIndex sheet. I see there are 1764 total artists listed there.

In [77]:
artists_artist_index = set(artists_index['Name'].dropna().str.strip())
len(artists_artist_index)

1764

Next, I'm going to create a set data structure, which ensures that all data items are unique and avoids any duplication. I'll then get the length of the set and do an initial inspection of its contents.

I can see there are 1764 artists in the artistsIndex sheet, and that while it's a good list, there are some data issues. For example, 

```
'Ballet Folklorico de San Antonio',
'Ballet Folkloricp de San Antonio',
```

or

```
'Billy Mata & the Texas Tradition',
'Billy Mata and the Texas Tradition',
```

All this suggests some manual cleanup may be helpful, but overall the collection is generally valid.

In [78]:
unique_artists = set()
unique_artists.update(artists_artist_index)

In [79]:
len(unique_artists)

1764

In [80]:
sorted(unique_artists)

['"The Italian Cowboy"',
 '(Fools?)',
 '(TLC - cancelled)',
 '01 Blues',
 '10 Years',
 '112',
 '12 Step Rebels',
 '13th Floor Elevators',
 '214th Army Band',
 '3 Balls of Fire',
 '4 2 MD',
 '4 on the Floor',
 '7seconds',
 '98 Degrees',
 'A Perfect Circle',
 'AC',
 'ANJ',
 'Aaron Barker',
 'Aaron Carter',
 'Accept',
 'Ace Frehley',
 'Adam Bomb',
 "Adler's Appetite",
 'Adolescents',
 'Adolph Hofner',
 'Adolph Hofner and the Pearl Wranglers',
 'Adrian y Destino',
 'Aerosmith',
 'Agustin Ramirez',
 'Air Force Concert Band',
 'Air Supply',
 'Airbourne',
 'Al Gomez',
 'Al Hirt w',
 'Al Olsen',
 'Al Sturchio Orchestra',
 'Al Sturchio and His Orchestra',
 'Alabama',
 'Alan Jackson',
 'Albert Collins and the Antone Blues Band',
 'Albert Garcia',
 'Albita',
 'Alcatrazz & Odin',
 'Alex Abravanel',
 'Alex Sutherlin Project',
 'Algiers Brass Band',
 'Ali Ekber Cicek',
 'Alibi',
 'Alice Cooper',
 'Alice Gomez',
 'All Time Low',
 'All-American Rejects',
 'Alondra',
 'Altered Natives',
 'Alvin Crow an

Now I'll look at the venue_artist_event sheet, which is the main one with all the concerts from the 1850s to the 2020s. There are 20320 unique artists in this list.

In [81]:
artists_venue_artist_event_sheet = set(venue_artist_event['Artist_Formula'].dropna().str.strip())

In [82]:
artists_venue_artist_event_sheet

{'Jamie DeAnda',
 'Natalie Rose',
 'Kolby Cooper',
 'Field of Seed',
 'Maenerchor',
 'The Texas All-Stars',
 'Stage 1 - Highway 101',
 'Cassidy',
 'The Junior Brown Band',
 'Ed Guymer and his Patriotic Band',
 'Thirteen',
 'Chick Corea Elektric Band',
 'Mateo Lopez',
 'Fifth Avenue Band',
 'Temblor del Norte',
 'Jimmy Witherspoon; Big Jim Winn and His Orchestra',
 'Pagame',
 'Golden Earring',
 'Sonny Ace',
 'Mariachi Campanas',
 'Small Faces',
 'Los Connyns',
 'Tailwind',
 'Momgoblin',
 'Santiago Jiminez',
 'Paul Jackson Jr.',
 'Frank Rodarte',
 'Blue Ridge Connection',
 'Joan Rivers',
 'Yelawolf',
 'J. Spockey',
 'Kelli King',
 'Roosevelt Sykes and His RCA Recording Orchestra',
 'La Nopalera',
 'Chase Matthew',
 'Chase Petra',
 'Alexander Ramirez',
 'Isidro Lopez',
 'Fuel',
 'Xceptthe1iwant',
 'Blackbirds',
 'Little Giant',
 'DJ Guero Vaquero',
 'the Shadow',
 'Forplay',
 'Earle Cobb School of Dance Revue',
 'Kenny Clark',
 'Rosario de la Fraga',
 'X-Greats',
 'Honorario Rodriguez',
 

In [83]:
len(artists_venue_artist_event_sheet)

20320

Finally, I look at the ven_ev_ar_address sheet. There are 11009 artists in this list, which focuses on the core years from 1970 to 2009. Over half the concerts in the data set are in these years.

In [84]:
artists_ven_ev_ar_address_sheet = set(ven_ev_ar_address['Artist'].dropna().str.strip())

In [85]:
len(artists_ven_ev_ar_address_sheet)

11009

Now, I'll update the unique artists set with the results from the two other sheets. By adding them to a set, duplicates will be eliminated. The length of the entire set is now 20419. 

### Investigate Artists Outside the Master Concert List

If there are 20419 total unique artists in the entire data set, but only 20320 in the master concert list, this means there are 99 artists somewhere that are not in the master concert sheet. Most likely these are artists that got typos edited in one sheet but not in another.

In [86]:
unique_artists.update(artists_venue_artist_event_sheet)

In [87]:
unique_artists.update(artists_ven_ev_ar_address_sheet)

In [88]:
len(unique_artists)

20419

To find out which are these 99 artists, I use a set difference operation. I verify it by checking the length of the resulting set. It is 99 artists. This suggests a need to investigate this particular group of names as they are inconsistent or problematic in some way.

In [89]:
only_in_artists_index = (artists_artist_index - artists_ven_ev_ar_address_sheet - artists_venue_artist_event_sheet)

In [90]:
only_in_artists_index

{'"The Italian Cowboy"',
 '(Fools?)',
 '(TLC - cancelled)',
 '01 Blues',
 '7seconds',
 'AC',
 'Al Hirt w',
 'Al Sturchio and His Orchestra',
 'Better than Ezra',
 'Bobby "Tossin\' and Turnin\'" Lewis',
 'Bobby Baker & the Longnecks',
 'Campanas De America,Ram Herrera',
 'Charlie and the Cool Cats',
 'Chick Corea (trio w',
 'DC',
 'Dave Wakeling',
 'Deborah Hay',
 'Earl "Fatha\' Hines',
 'Emerson',
 'Emotions',
 'English Beat w',
 'Evelyn "Champagne" King',
 'Frank Rodarte & los Blues Vatos',
 'Frank Rodarte & the Alamo All-Stars',
 'George Jones & Tammy Wynette',
 'George Strait,',
 'Hannah Montana',
 'Hate',
 'Henry Mancini w',
 'Infectious Grooves,',
 'Jim Cullum and Happy Jazz Band',
 'Jim Cullum and the Happy Jazz Band',
 "Jim Cullum's Happy Jazz Band",
 'Joe "King" Carrasco',
 'Joe "King" Carrasco and the Crowns',
 'Joe "King"Carrasco & Esteban Jordan',
 'Joe Reyes Quartet',
 'Joel Dilley Quintet',
 'Jr.',
 'Kiss',
 'LeAnn Rimes',
 'Less than Jake',
 'Lynn Anderson & Charlie Walke

In [91]:
len(only_in_artists_index)

99

### Create Artist Lookup Dictionary

Next, I'm going to create a lookup dictionary for artists with their index. This will make it very fast to look up whether an artist already has an index or not.

The rationale for this is as follows. We have 1769 listed artists with indexes already assigned. We don't want to reassign unique indexes; these may be in use somewhere already, so we want to retain unique indexes for all artists that already have one assigned. However, it is clear we have a lot more artists in the data than have these inexes assigned. So as we look over the unique list of all artists, we are going to want to look up whether it has an index, and if not assign it a new one. To make this an efficient and fast task, it's helpful to create a lookup dictionary in advance.

In [92]:
artist_to_index = dict(
    zip(
        artists_index["Name"],
        artists_index["Index"]
    )
)

### Test for Unique Artists in Ven_Ev_Ar_Address

I'd like the artists to be sortable chronologically in some way, so I'm going to add first appearance and last appearance information for them. To do this, I need to take the current text "Year", "Month", and "Day" columns and convert them into pandas datetime objects.

First, it occurs to me we may be able to save some time. The Ven_Ev_Ar_Address sheet was derived from the Venue_Artist_Event sheet, so I want to see if there are any aritsts in the first that are not in the second. I use a set difference operation. The result is an empty set. There are no artists in the first sheet that aren't in the second sheet. So I can do the date operation entirely on the second set.

In [93]:
artists_ven_ev_ar_address_sheet - artists_venue_artist_event_sheet

set()

### Add ISO Date for Concerts

Next, I'm going to use a utility I wrote some time ago that transforms string dates into ISO format date objects. Then I'll apply the method to every row in the venue_artist_event sheet and generate a new column, "date". Note that this is not writing the new column to the sheet, just adding it to the pandas data frame I'm working from.

In [94]:
import datetime
from dateutil import parser

def str_date_to_iso(month, day, year):
    
    str_date = f"{month} {day} {year}"
    dt = parser.parse(str_date).date().isoformat()
    
    return dt  

In [95]:
venue_artist_event["date"] = venue_artist_event.apply(lambda row: str_date_to_iso(row['Month'], row['Day'], row['Year']), axis=1)

When I ran the method, it flagged a concert in the data by Bob Wills on "February 29, 1962" at the Farmer's Daughter. There was no February 29 in 1962, so I changed it to February 28.

### Create Table With First and Last Artist Apperances

Next, I'm creating a new data frame with all the artists listed next to their first and last appearances in the concert list, by chronological order.

In [96]:
venue_artist_event_dates = (
    venue_artist_event
    .groupby("Artist_Formula")["date"]
    .agg(["min", "max"])
)

In [97]:
venue_artist_event_dates

,min,max
Artist_Formula,,
Jimmy Heath,1982-06-19,1982-06-19
Robert Casadesus (piano),1970-12-05,1970-12-05
!!! Chk Chk Chk,2023-10-12,2023-10-12
$Not,2023-05-09,2023-05-09
'Blue' Gene Tyranny,1985-06-28,1985-06-28
...,...,...
vocals],2010-02-27,2010-02-27
west,1986-02-08,1986-02-08
with added attraction Curley Mays,1958-03-06,1958-03-06


Inspecting this list, I can see a few issues straight away: an artist just called "vocals]', and an artist called "$Not", another one just called "west". These issues may bear further investigation. However, I will postpone data cleanup for now and proceed with organizing the data.

### Create Table With Unique Indexes for All Artists

In [98]:
artist_list_rows = []

next_index = artists_index["Index"].max() +1

for artist in unique_artists:
    
    if artist in artist_to_index:
        idx = artist_to_index[artist]
    else:
        idx = next_index
        next_index += 1

    if artist in venue_artist_event_dates.index:
        first = venue_artist_event_dates.loc[artist, "min"]
        last = venue_artist_event_dates.loc[artist, "max"]
    else:
        first = None
        last = None
        
    artist_list_rows.append({
        "Index": idx,
        "Artist": artist,
        "FirstAppearance": first,
        "LastAppearance": last,
    })

result = pd.DataFrame(artist_list_rows)

In [99]:
result

,Index,Artist,FirstAppearance,LastAppearance
0,1770,Jamie DeAnda,2022-04-02,2022-04-02
1,1771,Natalie Rose,2018-04-21,2023-08-27
2,1772,Kolby Cooper,2019-11-09,2023-07-08
3,1773,Field of Seed,2003-10-25,2003-10-25
4,1774,Blind Ignorents,1988-09-09,1988-09-09
...,...,...,...,...
20414,20421,December '63 (Franki Valli & the Four Seasons ...,2021-06-19,2021-06-19
20415,20422,Gordy Quist,2013-08-27,2013-08-27
20416,20423,The Bobby Boyle Trio (houseband),1979-09-12,1979-09-15
20417,20424,Hardluck Heroes,2007-09-02,2007-09-02


### Write Table to Excel

In [100]:
result.to_excel('sounds_data_to_share.xlsx', sheet_name='artistsIndex')

## Venues List

Now I'm going to turn my attention to the venues. As with the artists, we have different lists spread across different sheets. Some have coordinates, some do not. In this case we do not have a unique index for each venue anywhere in the data. We are going to want a list of all unique venues, with all available address and geodata, and a unique identifier for all of them.

We have a sheet called `Layer 1 Venues`. This list has 975 venues with addresses. It also has coordinates for 313 of these. 

Next there is a `Venues No Address` sheet, which I remember I derived at one point in order to be able to focus research on getting new coordinates. This sheet has 1080 venues on it. Considering that `Layer 1 Venues` has only 662 venues without coordinates, that leaves 418 venues in `Venues No Address` that can't be in `Layer 1 Venues`. This suggest some sort of data mismatch occurred, probably between the venues list in `Ven_Ev_Ar_Address` and the venues in `Layer 1 Venues`.

We have `venues_with_coordinates` in the revised product 1 master. This consists of the list from `Layer 1 Venues`, but with additional coordinates gathered by Angelina Cortes through R tidyverse. There are 965 venues with coordinates. One of them appears to have been subject to an editing error (The Copper Penny), because the longitude and latitude are the same except one has fewer decimals. There are ten venues still without coordinates in this list.

Then we have `Ven_Ev_Ar_Address`. This is the list of concerts between 1970 and 2009, which formed the basis for "Project 1" ("Get Lost") of Sounds of San Anto. 

Finally there is the authoritative source list, `Venue_Artist_Event`, with the complete list of all concerts and venues -- but no addresses or coordinates.

### Load and Inspect Data

I'll follow the same process as with artists. Firstly I'll develop unique lists based on each data source, then compare and form a unified list. I'll start by loading the sheets that I haven't yet loaded. Then I'll inspect the sheets that have venue information, so I can see the column headings.

In [231]:
layer_1_venues = pd.read_excel('spreadsheets/product1_data_master_rev.xlsx', sheet_name='Layer 1 Venues')
venues_no_address = pd.read_excel('spreadsheets/product1_data_master_rev.xlsx', sheet_name='Venues No Address')
venues_with_coordinates = pd.read_excel('spreadsheets/product1_data_master_rev.xlsx', sheet_name='venues_with_coordinates')

In [102]:
ven_ev_ar_address.head()

,Index,Month,Day,Year,Venue,Event_Artists,Address,Zip,Longitude,Latitude,Artist,Event,City,State
0,2581,January,1,1970,San Antonio Convention Center Arena,"KBER Holiday Show: Willie Nelson, Charley Pri...",601 Hemisfair Plaza Way,78205.0,NaN,NaN,Willie Nelson,KBER Holiday Show,San Antonio,Texas
1,2581,January,1,1970,San Antonio Convention Center Arena,"KBER Holiday Show: Willie Nelson, Charley Pri...",601 Hemisfair Plaza Way,78205.0,NaN,NaN,Charley Pride,KBER Holiday Show,San Antonio,Texas
2,2581,January,1,1970,San Antonio Convention Center Arena,"KBER Holiday Show: Willie Nelson, Charley Pri...",601 Hemisfair Plaza Way,78205.0,NaN,NaN,Grandpa Jones,KBER Holiday Show,San Antonio,Texas
3,2581,January,1,1970,San Antonio Convention Center Arena,"KBER Holiday Show: Willie Nelson, Charley Pri...",601 Hemisfair Plaza Way,78205.0,NaN,NaN,David Houston,KBER Holiday Show,San Antonio,Texas
4,2582,January,16,1970,San Antonio Municipal Auditorium,Vikki Carr,100 Auditorium Cir.,78205.0,NaN,NaN,Vikki Carr,NaN,San Antonio,Texas


In [103]:
venue_artist_event.head()

,Month,Day,Year,Venue,Event_Artists,Artist_Formula,Event_Formulas,Index,date
0,September,10,1857,St. Mary's Catholic Church,French Mountaineer Singers,French Mountaineer Singers,NaN,1,1857-09-10
1,June,25,1869,Menger Hotel,General Mason Farewell Party: San Antonio Bras...,San Antonio Brass Band,General Mason Farewell Party,2,1869-06-25
2,July,4,1874,San Pedro Park,San Antonio Brass Band,San Antonio Brass Band,NaN,3,1874-07-04
3,October,9,1874,"San Pedro Springs, Turner Hall",Tenth Texas Saengerfest,Tenth Texas Saengerfest,NaN,4,1874-10-09
4,October,10,1874,"Casino Hall, Turner Hall",Tenth Texas Saengerfest,Tenth Texas Saengerfest,NaN,5,1874-10-10


In [104]:
layer_1_venues.head()

,Venue_Name,Longitude,Latitude,Address,Zip,City,State
0,1033 Club,-98.480763,29.436116,1033 Avenue B,78215.0,San Antonio,Texas
1,151 Saloon,-98.690386,29.464097,10619 Westover Hills Blvd,78215.0,San Antonio,Texas
2,1902 Nightclub,-98.478216,29.421548,1174 East Commerce St,78205.0,San Antonio,Texas
3,210 Kapone's,-98.478815,29.425150,1223 East Houston,78205.0,San Antonio,Texas
4,4th Quarter Sports Bar,-98.569389,29.524564,8779 Wurzbach Rd,78240.0,San Antonio,Texas


In [105]:
venues_no_address.head()

,Venue
0,Bayt-El-Awhe Coffee House
1,Baptisit Temple Church
2,Providence Hall-OLLU
3,McCullom High School
4,Theater for the Perfoming Arts


In [106]:
venues_with_coordinates.head()

,Venue_Name,Longitude,Latitude,Address,Zip,City,State,...8,full_address,needs_geo,Unnamed: 10
0,1033 Club,-98.480763,29.436116,1033 Avenue B,78215.0,San Antonio,Texas,NaN,"1033 Avenue B, San Antonio, Texas, 78215",False,NaN
1,151 Saloon,-98.690386,29.464097,10619 Westover Hills Blvd,78215.0,San Antonio,Texas,NaN,"10619 Westover Hills Blvd, San Antonio, Texas,...",False,NaN
2,1902 Nightclub,-98.478216,29.421548,1174 East Commerce St,78205.0,San Antonio,Texas,NaN,"1174 East Commerce St, San Antonio, Texas, 78205",True,NaN
3,210 Kapone's,-98.478815,29.42515,1223 East Houston,78205.0,San Antonio,Texas,NaN,"1223 East Houston, San Antonio, Texas, 78205",True,NaN
4,4th Quarter Sports Bar,-98.569389,29.524564,8779 Wurzbach Rd,78240.0,San Antonio,Texas,NaN,"8779 Wurzbach Rd, San Antonio, Texas, 78240",False,NaN


The master sheet is venue_artist_event, so I'll start there. I get the list of unique individual venues listed there. There are 2602 in total.

In [107]:
venues_from_venue_artist_event = set(venue_artist_event['Venue'].dropna().str.strip())

In [108]:
len(venues_from_venue_artist_event)

2602

### Confirm Ven_Ev_Ar_Address is Subset of Venue_Artist_Event

Next, I'm going to look at the ven_ev_ar_address sheet. In theory, this is derived from venue_artist_event, only filtered by dates (1970-2009) and with geodata added. I'll derive the unique venues list from this sheet, then use a set difference operation to confirm this. The set of venues in ven_ev_ar_address that aren't also in venue_event_artist is an empty set, so we can effectively ignore this sheet in building a unique venues list.

In [109]:
venues_from_ven_ev_ar_address = set(ven_ev_ar_address['Venue'].dropna().str.strip())
len(venues_from_ven_ev_ar_address)

1548

In [110]:
venues_from_ven_ev_ar_address - venues_from_venue_artist_event

set()

### Investigate Layer 1 Venues and Analyze Differences With Other Venue Data

Next, I'll investigate the "Layer 1 Venues" sheet. This one does present some interesting results. It has 974 venues listed. When I use a set difference operator, I see there are 369 venues in Layer 1 Venues that are not in Venue_Artist_Event. I inspect the list, and find that some of these are spelling variants. As an example, in Layer 1 Venues we have the "Big Apple Night Club" but in Venue_Artist_Event this is "Big Apple Nite Club". Some of these are not so easily explained. Thus, we have the "Bier Garten Riverwalk" in Layer 1 Venues, but there is no mention of it in Venue_Artist_Event.

I conclude that these are somewhat separate lists. We have, on the one hand, a list of concerts with their associated venues, which is derived from the original SAPL librarian's list. Then there is this separate list of venues, of which about one third appear to be just names of venues, with no associated concert data in the other sheet.

The problem is, this (Layer 1 Venues) is the sheet we used as the primary basis for geocoding. But, it's not really the same as the list of concert locations. In retrospect, it would have been better to use the unique venues list from `ven_ev_ar_address` instead.

In [111]:
venues_from_layer_1_venues = set(layer_1_venues['Venue_Name'].dropna().str.strip())

In [112]:
in_layer_1_venues_but_not_ven_ar_ev = venues_from_layer_1_venues - venues_from_venue_artist_event

In [113]:
len(in_layer_1_venues_but_not_ven_ar_ev)

369

In [114]:
in_layer_1_venues_but_not_ven_ar_ev

{'108 Noth Centre',
 '4th Quarter Sports Bar',
 'Abracadabra',
 'Alamo Ice House',
 "Aldaco's Mexican Restaurant",
 "Aldo's Ristorante Italian",
 'Antonio Ballroom',
 'Antonio Strad Violin',
 'Apple Tree Lounge',
 'Aquariaum Disco',
 'Aragon Blue Room',
 'Artisan on Alamo Distillery',
 'Artpace San Antonio',
 'Azuca Nuevo Latino Restaurant & Bar',
 'Bar 414',
 'Bar America',
 'Bar Rojo at The Grand Hyatt Hotel',
 "Barclay's Piano Bar",
 "Beto's Comida Latina",
 'Bier Garten Riverwalk',
 'Big Apple Night Spot',
 'Big Band Fun Club',
 "Big Bob's Burgers",
 'Big Texas Ice House',
 'Bihl Haus Arts',
 "Billy D's Club",
 'Black Canteen',
 'Blind Tiger Comedy Club',
 'Blue Note Lounge',
 'Blue Star Brewing Co.',
 "Bo's Rodeo",
 "Bohanan's Prime Steaks & Seafood",
 'Boiler House Texas Grill & Wine Garden',
 'Bombay Bicycle Club',
 'Bone Club',
 'Bonesshakers Tap House and Pizzeria',
 'Boozehounds',
 'Botika Peruvian-Asian Restaurant',
 'Bottom Bracket Social Club',
 'Boudro’s Texas Bistro',
 '

In [115]:
len(layer_1_venues)

974

### Analyze Venues_With_Coordinates and Differences With Layer 1 Venues

Next, I look at venues_with_coordinates. This should be identical to layer 1 venues, but it isn't. There are 18 venues that appear in Layer 1 Venues but not in venues_with_coordinates. This should be impossible. Closer inspection suggests these are cases where symbols have been differently represented, possibly due to software (Microsoft Excel?) or operating system (Word vs. Mac) differences.

For example, the dash in 'The Carver Community Cultural Center – The Jo Long Theatre for Performing Arts' is represented as '–' in venues_with_coordinates. I can't tell why Python thinks there is a difference between the representations of 'Raul's``'. I do know those backticks shouldn't be there. In venues_with_coordinates, acceents have been transposed, for example Rosario's Café y Cantina is Rosario's Café y Cantina. 

Others appear to have been corrected in venues_with_coordinates. For example, '108 Noth Centre', clearly a typo, has been changed to '108 North Centre' in venues_with_coordinates.

So, these 18 venues will generate false double listings in our eventual unique venues list and require manual editing and attention.

In [116]:
venues_from_venues_with_coordinates = set(venues_with_coordinates['Venue_Name'].dropna().str.strip())
venues_from_layer_1_venues - venues_from_venues_with_coordinates

{'"Menudo Acres"',
 '108 Noth Centre',
 "Alexander's Bar and Café",
 'Bonesshakers Tap House and Pizzeria',
 'Boudro’s Texas Bistro',
 'Fralo’s Art of Pizza',
 'Maggiano’s Little Italy San Antonio',
 "Mary's Café",
 'Perry’s Steakhouse and Grille',
 'Powerhouse Café',
 "Raul's``",
 "Rosario's Café y Cantina",
 'Roxy Sports Bar -2801 Pat Booker Rd, Universal City, TX 78148',
 'Ruth’s Chris Steak House –La Cantera',
 'Teddy Bear Lounge & Café',
 'The Carver Community Cultural Center – The Jo Long Theatre for Performing Arts',
 'The Frutería -Botanero',
 'Web House Café & Bar'}

In [117]:
venues_from_venues_with_coordinates

{'1033 Club',
 '108 North Centre',
 '1428 S. Presa St.',
 '151 Saloon',
 '1902 Nightclub',
 "210 Kapone's",
 '321 Max',
 '4th Quarter Sports Bar',
 '500 Block Madison',
 '500 Carolina Street',
 '502 Bar',
 '801 West Russell Place',
 'AT&T Center',
 'Abracadabra',
 "Acapulco Sam's",
 'Accents',
 'Acquarium III',
 'After Midnight',
 'Aggie Park',
 'Agudas Achim Synagogue',
 'Alameda Theater',
 'Alamo Beer Company',
 'Alamo City Music Hall',
 'Alamo Heights United Methodist Church',
 'Alamo Ice House',
 'Alamo Plaza',
 'Alamodome',
 "Aldaco's Mexican Restaurant",
 "Aldo's Ristorante Italian",
 "Alexander's Bar and Café",
 "Alfred's",
 'Alzafar Shrine Temple',
 'American Legion Post 539',
 'Anaqua Room',
 "Andrew's 281 Live",
 'Anhalt Hall',
 "Annie & Ed's",
 'Antonian Ballroom',
 'Antonio Ballroom',
 'Antonio Strad Violin',
 'Apple Tree Lounge',
 'Aquariaum Disco',
 'Aragon Blue Room',
 "Arjon's International Club",
 'Arneson River Theatre',
 "Arthur's",
 'Artisan on Alamo Distillery',
 

We have some other clear inconsistencies visible on inspection, such as

```
'Repitlez Sports Bar',
'Reptilez Live Music Venue',
```

So, we will need a bit of human TLC here.

### Investigate Source of Venues No Address Sheet

The question remains how and from where I derived the `Venues No Address` sheet. It is clear that this was at the start of summer 2025, when the issue of venues without coordinates was raised and work was begun on resolving it. 

In [118]:
len(venues_from_ven_ev_ar_address - venues_from_layer_1_venues)

1079

Here is the answer. If we do the set difference between all the venues in `ven_ev_ar_address` and all the venues in `layer 1 venues` we find that there are 1079 venues in `ven_ev_ar_address` that are *not* in `layer 1 venues`. The number of rows in `Venues No Address` is 1079. This makes it clear that what I did is look up venues in `ven_ev_ar_address` that had no address. The venues that do have an address on this list are all derived from `Layer 1 Venues`. In other words, originally venue addresses and coordinates were placed on the Layer 1 Venues sheet, and wherever there was a match in `ven_ev_ar_address`, imported there. All the venues in `ven_ev_ar_address` that are not in `Layer 1 Venues` have no address.

This might have improved subsequently. Originally, as we have seen, only 313 venues in `Layer 1 Venues` had addresses; this was subsequently raised to 964, and the resulting new venue coordinates were imported directly to the Redis database that underlies the Sounds of San Anto "Get Lost" project. But the spreadsheet was never altered; the new venue coordinates live only in the `venues_with_coordinates` sheet.

I'll test this hypothesis by finding out how many venues in `ven_ev_ar_address` have no address. First, I'll find out the exact column names in this data frame. Then I'll created a filtered data frame with just the entries that have no address. Then I'll get the list of unique venues in that list. It is 1078 venues. So that is definitely the source of my "Venues No Address" sheet: a search of `ven_ev_ar_address` for all venues with no address.

In [119]:
ven_ev_ar_address.columns

Index(['Index', 'Month', 'Day', 'Year', 'Venue', 'Event_Artists', 'Address',
       'Zip', 'Longitude', 'Latitude', 'Artist', 'Event', 'City', 'State'],
      dtype='str')

In [120]:
missing_address = ven_ev_ar_address[
    ven_ev_ar_address["Address"].isna() 
]

missing_address = set(sorted(missing_address["Venue"].dropna().unique()))

print(f"{len(missing_address)} venues with no address:")
print(missing_address)

1078 venues with no address:
{'Los Generales', 'San Pedro Park', 'Sin 13', 'Fourth Street Inn', 'Loftin Student Center-SAC', 'Crickets', 'SeaWorld', 'Make My Day', 'La Cantera Golf Club', 'OrderUp (Lincoln Heights)', 'Thunderbird Hills Free United Methodist Church', 'Armadillo on the Creek (Comfort)', 'Brooklyn Bar and Grill', 'Knights of Columbus Hall (Seguin)', 'Madison High School', 'The Quarry', 'Texas Pride Barbecue', 'La Tuna', 'Blue Bubble Ballroom', 'St. Paul Drugstore Community Cafe', 'Atomix', 'Saengerhalle (New Braunfels)', 'Colonial Hills UMC', 'Cow Town Dance Hall (Lytle)', 'Life Christian Center (New Braunfels)', 'Texas Star Ballroom', 'Chihuahua Street between Sabinas and Trinity', "Big Daddy's (Bulverde)", 'Countryman Kitchen', 'Cattleman Square Tavern', 'Juarez Plaza', 'Prime Steakhouse (Kirby)', 'Republic of Texas Plaza', 'El Rey Perido Night Club', "Crumpet's", 'Red Hot & Blues', 'New Hope Baptist Church', 'SAC', 'Shangri-La Chinese Restaurant', "Sumthin Dif'rent", '

### Investigate Venues With No Coordinates

However, it's not that simple. Our issues run a bit deeper than that. There are venues that have an address but have no coordinates. If we run the same analysis, but look for venues that lack coordinates, we find 1471 venues. There are 393 venues that have an address but no coordinates.

We should also keep in mind that, when, eventually (December 2025) the R tidyverse library was used to add missing address and coordinate data, it was based on τhe Layer 1 Venues sheet, and not from the Venues No Address sheet I had generated from `ven_ev_ar_address`. Even if that were not the case, I had not, at that time, realized we had an additional 393 venues with an address but no coordinates. So, it will be worthwhile now to unify all our data on venues, addresses, and coordinates and get a unified count of all missing and incomplete geolocation data.

In [121]:
missing_coords = ven_ev_ar_address[
    ven_ev_ar_address["Longitude"].isna() | ven_ev_ar_address["Latitude"].isna()
]

missing_coords = set(sorted(missing_coords["Venue"].dropna().unique()))

print(f"{len(missing_coords)} venues with no coordinates:")
print(missing_coords)

1471 venues with no coordinates:
{'Club Abrazo', 'Make My Day', 'Saengerhalle (New Braunfels)', 'Chihuahua Street between Sabinas and Trinity', 'Juarez Plaza', 'Republic of Texas Plaza', "Panchito's", 'MacArthur Field, Ft. Sam Houston', 'La Hacienda', 'Third Wave', "Clay's Ice House", 'The Nugget', 'Lubianski Dance Hall', 'San Antontio Events Center', 'Pecan Grove', "St. Luke's Episcopal Church", 'Olmos Basin Park', 'Naked Iguana Tequila Bar', 'Carmen de La Calle', 'Pump House', 'The Inn at Turtle Creek', 'Fort Sam Houston Polo Grounds', 'Villareal Outdoor Concert Site (Helotes)', 'Club Miami', "Myrtle's", 'Bandera Road Community Church', 'Red Hot and Blues', 'St. Francis Episcopal Church', 'Intituto de Mexico', 'Ajuua Cuisine de Mexico', 'Steamers', 'Hollywood Café', 'Finish Line (Jourdanton)', "Terry's Lounge", 'The Warehouse Music Venue', 'Instituto Cultural Mexicano', 'Earl of Country', 'Imperial Ballroom', 'Wal Mart (SE Military)', 'Station Break', 'Sting Lounge', "Arturo's Sports

### Investigate Venues With Address But No Coordinates

In [122]:
missing_coords - missing_address

{'"Menudo Acres"',
 '1902 Nightclub',
 '321 Max',
 "Acapulco Sam's",
 'Accents',
 'Acquarium III',
 'After Midnight',
 'Alameda Theater',
 'Alamo Heights United Methodist Church',
 'Alamo Plaza',
 "Alexander's Bar and Café",
 "Alfred's",
 'Alzafar Shrine Temple',
 'American Legion Post 539',
 'Anaqua Room',
 "Andrew's 281 Live",
 "Annie & Ed's",
 'Antonian Ballroom',
 "Arthur's",
 "Arturo's Sports Bar & Grill",
 'Austin Municipal Auditorium',
 "BJ's Night Club",
 'Backstage Live',
 'Badlands Saloon',
 'Bamboo Room',
 'Baptist Temple Church',
 'Billy Blues BBQ',
 'Billy Marbles',
 'Black Orchid',
 'Blossom Athletic Center',
 'Boardwalk Bistro',
 "Bogart's",
 "Bonnie Jean's",
 'Bottom Line',
 'Brackenridge Park',
 "Brandi's Piano Bar",
 "Brazakka's Reef",
 'Broad Street',
 "Broadway Turk's",
 'Brooklyn Street Bar and Grill',
 "Buckley's Jazz Bar",
 "Bucko's Boardroom",
 "Buddy's Ize House",
 "Budro's",
 'Burro Land',
 'Bwana Dik',
 'Cabaret',
 'Cactus Beer Garden',
 'Cafe Americain',
 'C

In [123]:
len(missing_coords - missing_address)

393

### Build Unique Venues List

We're ready to construct our unique venue list. Let's go ahead and fuse our various sources.

In [124]:
unique_venues = venues_from_venue_artist_event | venues_from_layer_1_venues | venues_from_venues_with_coordinates

In [125]:
len(unique_venues)

2988

We have a total of 2988 venues. These are 2602 venues from the original `venue_artist_event` sheet, another 369 that are unique to `layer 1 venues`, and an additional 17 that are duplicates with alternative formatting or spelling from `venues_with_coordinates`.

### Ensure Data Frame Has ISO Date Column

I will double-check that my venue_artist_event data frame still has a date column. If the notebook has shut down and not been fully run again, that column may be gone. If it is gone, I'll add it again.

In [126]:
venue_artist_event.columns

Index(['Month', 'Day', 'Year', 'Venue', 'Event_Artists', 'Artist_Formula',
       'Event_Formulas', 'Index', 'date'],
      dtype='str')

In [143]:
## venue_artist_event["date"] = venue_artist_event.apply(lambda row: str_date_to_iso(row['Month'], row['Day'], row['Year']), axis=1)

In [144]:
## venue_artist_event.columns

### Add First and Last Appearance Dates for Venues

As in the case of artists, I would like the first and last appearance dates of any given venue. On this occasion, I would even be interested to order the venue indexes that way, with the oldest-appearing venues indexed first, and the latest-appearing indexed last. We do have some venues with no concerts at all, listed out in Layer 1 Venues, those will come at the very end.

I'll get that information on first and last appearances, then sort by order of first appearance.

In [129]:
venue_dates = (
    venue_artist_event
    .groupby("Venue")["date"]
    .agg(
        first_appearance="min",
        last_appearance="max"
    )
    .reset_index()
)

In [130]:
venue_dates.head()

,Venue,first_appearance,last_appearance
0,"""Menudo Acres""",1978-10-06,1978-10-07
1,1033 Club,1991-06-24,1991-09-28
2,110 Broadway Building Office,1985-07-12,1987-08-10
3,11th Street Cowboy Bar (Bandera),2007-07-14,2023-07-29
4,136 South Acme,2023-03-04,2023-03-04


In [131]:
venue_dates = venue_dates.sort_values(
    by="first_appearance"
).reset_index(drop=True)

In [132]:
venue_dates.head(40)

,Venue,first_appearance,last_appearance
0,St. Mary's Catholic Church,1857-09-10,1857-09-10
1,Menger Hotel,1869-06-25,1980-01-26
2,San Pedro Park,1874-07-04,1991-11-02
3,"San Pedro Springs, Turner Hall",1874-10-09,1874-10-09
4,"Casino Hall, Turner Hall",1874-10-10,1874-10-10
5,"San Pedro Springs, Alamo Plaza, Casino Hall",1874-10-11,1874-10-11
6,Travis Park,1883-01-24,2021-09-25
7,Casino Hall,1883-01-31,1908-05-27
8,Turner Hall,1883-02-01,1900-04-19
9,San Pedro Springs Park,1883-03-24,1883-04-01


### Add Unique Venues Indexes

Next, we'll give them unique indexes.

In [133]:
venue_dates.insert(
    0,
    "Index",
    range(1, len(venue_dates) + 1)
)

In [134]:
venue_dates.head()

,Index,Venue,first_appearance,last_appearance
0,1,St. Mary's Catholic Church,1857-09-10,1857-09-10
1,2,Menger Hotel,1869-06-25,1980-01-26
2,3,San Pedro Park,1874-07-04,1991-11-02
3,4,"San Pedro Springs, Turner Hall",1874-10-09,1874-10-09
4,5,"Casino Hall, Turner Hall",1874-10-10,1874-10-10


The next thing we need to do is add the venues that have no concert dates associated with them. These are all located in the `Layer 1 Venues` and `venues_with_coordinates` sheets.

Earlier, I developed a set of all the unique venues, called `unique_venues`, and just now we got the complete list of all the venues that do have concert data associated. So we need to find the difference. I'll derive the set of venues that do have concerts as follows:

In [135]:
venues_with_concerts = set(venue_dates["Venue"])

Now I'll do a set difference between the complete list of venues and this list.

In [136]:
venues_no_concerts = sorted(unique_venues - venues_with_concerts)

In [137]:
len(venues_no_concerts)

386

To add these to the end of the current data frame, I will create a new data frame and then concatenate the two. This strategy uses exclusively pandas rather than regular Python code, and since pandas leverages the C language this is very fast and computationally efficient.

In [138]:
start = len(venue_dates) + 1

new_data = pd.DataFrame({
    "Index": range(start, start + len(venues_no_concerts)),
    "Venue": venues_no_concerts,
    "first_appearance": pd.NaT,
    "last_appearance": pd.NaT
})

In [139]:
all_venues_df = pd.concat(
    [venue_dates, new_data],
    ignore_index=True
)

In [140]:
all_venues_df.tail(50)

,Index,Venue,first_appearance,last_appearance
2939,2940,The Other Women,NaT,NaT
2940,2941,The Phantom Room,NaT,NaT
2941,2942,The Ringer Pub,NaT,NaT
2942,2943,The Rock,NaT,NaT
2943,2944,The Rockin' K,NaT,NaT
2944,2945,The Service Station Cafe and Beer Garden,NaT,NaT
2945,2946,The Settlement,NaT,NaT
2946,2947,The Silver Slipper Nite Club,NaT,NaT
2947,2948,The Squeezebox,NaT,NaT
2948,2949,The Three-Legged Monkey,NaT,NaT


### Write Venues Data Frame to Excel

And now we can take this information and write it to our Excel file as a new sheet.

In [142]:
with pd.ExcelWriter(
    "sounds_data_to_share.xlsx",
    engine="openpyxl",
    mode="a",
    if_sheet_exists="replace"
) as writer:
    all_venues_df.to_excel(
        writer,
        sheet_name="VenuesIndex",
        index=False
    )

## Venue Geodata

The next task is to add the most current geodata for all venues. I'll want to investigate what venues have geodata, identify any inconsistencies or problems with the existing data, and compile a list of venues that still have no data.

Venue geodata is located in the following places: `Layer 1 Venues`, `Ven_Ev_Ar_Address`, and `venues_with_coordinates`. In addition, there are two pdf tables generated by a student worker with a few venues investigated by hand. The first thing I will do is try to work the student's data into Python, so all the data is in a suitable format for comparison.

### Investigating the student's Data

The student worked off a spreadsheet based on `Layer 1 Venues`. At the start of the summer of 2025, I was asked to identify the venues that did not have coordinates. I ran a series of analyses on the data, querying not just missing coordinates, but also some other venues with issues, such as venues with missing or bad addresses, venues with missing zip codes, and venues with coordinates that placed them over 30 miles from the center of San Antonio. All these lists eventually found their way to a spreadsheet called `venues_separated.xlsx`, on 6-2-25, which I shared with Diane Lopez, and through her with the student worker. 

It is unfortunate that at that time, I had not understood the difference between `Layer 1 Venues` and `ven_ev_ar_address`, nor that it was this latter that really needed attention. As a result, work was done predominantly on `Layer 1 Venues`, which was missing some 600 or so coordinates, and not on `ven_ev_ar_address`, which was in worse shape with over 1400 missing coordinates.

Eventually, in January 2026, the student worker shared a list of 58 venues derived from the `venues_separated` document, most of them with some sort of address and/or coordinates, and some with comments. These were shared in a pdf table called `venues_separated ZIP.pdf`. There was a comment that a tool had been used for zip code lookup: [zip tool](https://tools.usps.com/zip-code-lookup.htm?byaddress).

The pdf had text that could be highlighed and copied, but not readily converted to Excel readable format, so I used Claude AI to generate a spreadsheet, which it named `sa_venues.xslx`.

In [284]:
student_data = pd.read_excel('spreadsheets/sa_venues.xlsx', sheet_name='Venues')

In [285]:
student_data

,ID,NAME,ADDRESS,CITY,STATE,ZIP,LONGITUDE,LATITUDE,NOTES
0,320,Ivory Lounge & Nightclub,1201 Austin Highway,San Antonio,Texas,78209.0,-98.4468,29.48944,zip tool: https://tools.usps.com/zip-code-lookup.htm?byaddress
1,347,Just Country,9323 Perrin Beitel,San Antonio,Texas,78217.0,-94.6424,30.14632,NaN
2,352,Kelly Field Terrace,corner Mission and Mitchell Road,San Antonio,Texas,78214.0,-100.8720,32.40086,no specific address
3,357,Keystone Club,5152 Fredericksburg,San Antonio,Texas,78229.0,-98.8720,30.27520,NaN
4,366,Krueger's Bar and Party House,6870 East Highway 87,San Antonio,Texas,78263.0,-94.6424,30.14632,NaN
5,373,La Rumba Nite Club,411 Old Hwy 90,San Antonio,Texas,78237.0,-96.5625,29.70536,NaN
6,381,Larry's Cantina,10929 Nacogdoches,San Antonio,Texas,78217.0,NaN,NaN,NaN
7,382,Larry's Lounge,2219 South Flores,San Antonio,Texas,78204.0,NaN,NaN,NaN
8,383,Lartos Lounge,2303 East Commerce,San Antonio,Texas,78203.0,-95.9000,33.24713,NaN
9,388,Last National Bank,10127 Coachlight,San Antonio,Texas,78216.0,NaN,NaN,NaN


In [286]:
len(student_data)

58

In all, this data set contributes 52 new coordinates (six remained unresolved) out of the 71 that were listed as missing coordinates in the `venues_separated` spreadsheet. 

In investigating the history of this data, and tracking down how I originally created the `venues_separated` spreadsheet, I found this Python script, dated 6-2-2025.

In [156]:
# import requests
# import pandas as pd
# from math import radians, cos, sin, asin, sqrt

# # API Setup
# express_address = "dev.cedish.utsa.edu/devconcerts-api"
# protocol = "https"
# API_ADDRESS = f"{protocol}://{express_address}"

# # San Antonio coordinates (center point)
# SA_LAT = 29.4241
# SA_LON = -98.4936

# def fetch_data(key):
#     try:
#         url = f"{API_ADDRESS}/json/{key}"
#         response = requests.get(url)
#         response.raise_for_status()
#         return response.json()
#     except requests.exceptions.RequestException as e:
#         print("Problem with fetch operation:", e)
#         return None

# def haversine(lon1, lat1, lon2, lat2):
#     # Convert decimal degrees to radians
#     lon1, lat1, lon2, lat2 = map(radians, [lon1, lat1, lon2, lat2])
#     # Haversine formula
#     dlon = lon2 - lon1
#     dlat = lat2 - lat1
#     a = sin(dlat / 2) ** 2 + cos(lat1) * cos(lat2) * sin(dlon / 2) ** 2
#     c = 2 * asin(sqrt(a))
#     r = 3958.8  # Radius of Earth in miles
#     return c * r

# # Fetch venues
# data = fetch_data('venues')

# if data and 'venues' in data:
#     df = pd.DataFrame(data['venues'])

#     # Missing coordinates
#     missing_coords_df = df[df['latitude'].isnull() | df['longitude'].isnull()]

#     # Far from San Antonio
#     df_with_coords = df.dropna(subset=['latitude', 'longitude']).copy()
#     df_with_coords['distance_from_SA'] = df_with_coords.apply(
#         lambda row: haversine(SA_LON, SA_LAT, row['longitude'], row['latitude']), axis=1
#     )
#     far_df = df_with_coords[df_with_coords['distance_from_SA'] > 30]

#     # Export to Excel with three sheets
#     with pd.ExcelWriter("venues_separated.xlsx", engine='openpyxl') as writer:
#         df.to_excel(writer, sheet_name='All Venues', index=False)
#         missing_coords_df.to_excel(writer, sheet_name='Missing Coordinates', index=False)
#         far_df.to_excel(writer, sheet_name='>30 Miles Away', index=False)

#     print("Exported to venues_separated.xlsx")
# else:
#     print("No venue data to export.")

This script shows that in order to generate the sheet that Marisela and Diane worked from, I directly queried the Redis database venues list. The result was a list 974 items long, matching Layer 1 Venues. This indicates that I built the venues list for project 1 from Layer 1 Venues directly. Right now, I'll just point that out, and we'll see later if this becomes significant. I suspect it will, since the venues list in `Layer 1 Venues` is different from the venues present in the concert lists.

A second pdf that the student worker shared had 36 venues. This appears to be work done on the "Missing or Bad Addresses" sheet from the `venues_separated` spreadsheet. Some of the locations have had addresses revised and notes made on them. This pdf was entitled `venues_separated MISSING OR BAD ADDRESSES.pdf`. Again, I asked Claude AI to generate an excel spreadsheet from the copy-pasted text. The result was a spreadsheet called `sa_venues_2.xlsx`.

In [287]:
student_address_revisions = pd.read_excel("spreadsheets/sa_venues_2.xlsx", sheet_name="Venues")

In [288]:
student_address_revisions

,ID,NAME,ADDRESS,CITY,STATE,ZIP,LONGITUDE,LATITUDE,NOTES
0,7.0,Abracadabra,258 Central Park Mall,San Antonio,Texas,78216.0,-98.4951,29.42460,NaN
1,36.0,Avalon Grill,509 1/2 E Commerce St,San Antonio,Texas,78205.0,-98.4951,29.42460,NaN
2,53.0,Big Bob's Burgers,2215 Harry Wurzbach Road,San Antonio,Texas,78212.0,-98.5113,29.46646,NaN
3,80.0,Bo's Rodeo,Hwy 173,Bandera,Texas,78003.0,-99.0725,29.72701,NaN
4,81.0,Botika Peruvian-Asian Restaurant,"303 Pearl Parkway, Suite 111",San Antonio,Texas,78215.0,-98.4800,29.44250,NaN
5,126.0,Casey's Saloon and Dance,West Crockett Street,San Antonio,Texas,78205.0,-98.4885,29.42426,slight adjustment to address; most recent reference 2022 article; don't have suite number
6,127.0,Casino Club,102 W Crockett St.,San Antonio,Texas,78205.0,-98.5249,29.44540,NaN
7,NaN,Edge of Town,Culebra and Ingram,San Antonio,Texas,78251.0,-98.6658,29.45796,found articles discussing 'edge of town' around bandera area. Need to clarify if this is an actual establishment or general area.
8,NaN,Engle's Place,Nacogdoches Rd,San Antonio,Texas,78217.0,-98.4951,29.42460,Difficult find
9,352.0,Kelly Field Terrace,corner Mission and Mitchell Road,NaN,Texas,78214.0,-100.8720,32.40086,found zip code; no specific address


This data contributes 35 address revisions to the data set.

I notice that on its way from pdf to Excel to pandas, the ids and zip codes have picked up decimal points. To deal with that, we're going to want to recast those columns as strings. 

In attempting to do that I notice that some of the venues (Edge of Town and Engle's Place) are missing IDs. I'll restore them based on their location in the Layer 1 Venues sheet. We won't need these IDs, since we've assigned all-new indexes to the entire unique venues list using a different formula, but it's not good to have bad data anywhere.

In [289]:
student_address_revisions["ID"] = (
    student_address_revisions["ID"]
    .apply(lambda x: str(int(x)) if pd.notna(x) else "")
)

student_address_revisions["ZIP"] = (
    student_address_revisions["ZIP"]
    .apply(lambda x: str(int(x)) if pd.notna(x) else "")
)

In [290]:
student_address_revisions.at[7, "ID"] = "218"
student_address_revisions.at[8, "ID"] = "236"

In [291]:
student_address_revisions.head(10)

,ID,NAME,ADDRESS,CITY,STATE,ZIP,LONGITUDE,LATITUDE,NOTES
0,7,Abracadabra,258 Central Park Mall,San Antonio,Texas,78216,-98.4951,29.42460,NaN
1,36,Avalon Grill,509 1/2 E Commerce St,San Antonio,Texas,78205,-98.4951,29.42460,NaN
2,53,Big Bob's Burgers,2215 Harry Wurzbach Road,San Antonio,Texas,78212,-98.5113,29.46646,NaN
3,80,Bo's Rodeo,Hwy 173,Bandera,Texas,78003,-99.0725,29.72701,NaN
4,81,Botika Peruvian-Asian Restaurant,"303 Pearl Parkway, Suite 111",San Antonio,Texas,78215,-98.4800,29.44250,NaN
5,126,Casey's Saloon and Dance,West Crockett Street,San Antonio,Texas,78205,-98.4885,29.42426,slight adjustment to address; most recent reference 2022 article; don't have suite number
6,127,Casino Club,102 W Crockett St.,San Antonio,Texas,78205,-98.5249,29.44540,NaN
7,218,Edge of Town,Culebra and Ingram,San Antonio,Texas,78251,-98.6658,29.45796,found articles discussing 'edge of town' around bandera area. Need to clarify if this is an actual establishment or general area.
8,236,Engle's Place,Nacogdoches Rd,San Antonio,Texas,78217,-98.4951,29.42460,Difficult find
9,352,Kelly Field Terrace,corner Mission and Mitchell Road,NaN,Texas,78214,-100.8720,32.40086,found zip code; no specific address


I'm going to apply the same technique to the first data frame, where zip codes have also been rendered as numbers with a decimal point.

In [292]:
student_data["ZIP"] = (
    student_data["ZIP"]
    .apply(lambda x: str(int(x)) if pd.notna(x) else "")
)

In [293]:
student_data.head()

,ID,NAME,ADDRESS,CITY,STATE,ZIP,LONGITUDE,LATITUDE,NOTES
0,320,Ivory Lounge & Nightclub,1201 Austin Highway,San Antonio,Texas,78209,-98.4468,29.48944,zip tool: https://tools.usps.com/zip-code-lookup.htm?byaddress
1,347,Just Country,9323 Perrin Beitel,San Antonio,Texas,78217,-94.6424,30.14632,NaN
2,352,Kelly Field Terrace,corner Mission and Mitchell Road,San Antonio,Texas,78214,-100.8720,32.40086,no specific address
3,357,Keystone Club,5152 Fredericksburg,San Antonio,Texas,78229,-98.8720,30.27520,NaN
4,366,Krueger's Bar and Party House,6870 East Highway 87,San Antonio,Texas,78263,-94.6424,30.14632,NaN


### Reflections on the Development Process

Something is striking me about this information, and that is, that as of June 2025, when I queried the Redis database, there were only 71 venues missing coordinates in the `venues` JSON. Going back through previous versions of `venues.json` that I have stored on my computer, this was true as far back as March 2024, soon after I arrived, and after I queried the Mapbox API for data. A `venues.json` file from March 2024 has 974 venues and 72 venues missing longitude and latitude data. It looks like after querying the API and updating the venues JSON, there were only about 7% of venues missing data (71/974). However, the master spreadsheet was never updated in the Layer 1 Venues layer, leading to a mismatch between the spreadsheet and the application database. The spreadsheet continued to show 662 out of 974 missing coordinates. Furthermore, this is also the case with the `years.json` that underlay the display of concerts by years. Here, the problem was even less, with only 41 venues missing coordinates. This is because some of the venues in 'Layer 1 Venues' don't actually have concerts associated with them, so not all were represented in the concerts.

This suggests the missing venues problem was a bit misconstrued, and when some venues were that were showing up in the concert list panel were not showing up in the map as markers, the problem may have been other than the underlying data. When the problem was addressed in January 2026, it was fixed. At the moment, the main issue is not to analyze what went wrong during the development process, but to ensure that the data is correct for publication. But, it is clear that failure to update the spreadsheet after an API query led to a misunderstanding of what the actual gaps were in our data, led to us thinking there were  more venues missing coordinates than there actually were, and probably to a misdiagnosis of the problem and unnecessarily duplicated work. Specifically, when the 'Layer 1 Venues' list was given to Angelina Cortes, it was from the spreadsheet, showing 662 missing coordinates; when data was given to the student worker, it was straight from a Redis database query, showing only about 71 venues missing data. 

All this suggests in the future it is absolutely essential to have a clearly understood source of truth for all data that is constantly and transparently updated.

### Load and Inspect Angelina's data

After the student worked on the venues missing coordinates or having bad addresses, given the visual mismatch between concert venues in the concerts panel and on the map, and that based on the Layer 1 Venues worksheet of the product 1 master spreadsheet, it looked like 662 venues were still missing coordinates, Angelina Cortes directed the R library tidyverse to retrieve coordinate data, resulting in a worksheet called `venues_with_coordinates`.

I'll now load Angelina's data, examine it, and since the Zip is an integer, correct it to a string as with the other data sets.

In [198]:
angelina_data = pd.read_excel('spreadsheets/product1_data_master_rev.xlsx', sheet_name='venues_with_coordinates')

In [199]:
angelina_data.head()

,Venue_Name,Longitude,Latitude,Address,Zip,City,State,...8,full_address,needs_geo,Unnamed: 10
0,1033 Club,-98.480763,29.436116,1033 Avenue B,78215.0,San Antonio,Texas,NaN,"1033 Avenue B, San Antonio, Texas, 78215",False,NaN
1,151 Saloon,-98.690386,29.464097,10619 Westover Hills Blvd,78215.0,San Antonio,Texas,NaN,"10619 Westover Hills Blvd, San Antonio, Texas,...",False,NaN
2,1902 Nightclub,-98.478216,29.421548,1174 East Commerce St,78205.0,San Antonio,Texas,NaN,"1174 East Commerce St, San Antonio, Texas, 78205",True,NaN
3,210 Kapone's,-98.478815,29.42515,1223 East Houston,78205.0,San Antonio,Texas,NaN,"1223 East Houston, San Antonio, Texas, 78205",True,NaN
4,4th Quarter Sports Bar,-98.569389,29.524564,8779 Wurzbach Rd,78240.0,San Antonio,Texas,NaN,"8779 Wurzbach Rd, San Antonio, Texas, 78240",False,NaN


In [200]:
angelina_data.columns

Index(['Venue_Name', 'Longitude', 'Latitude', 'Address', 'Zip', 'City',
       'State', '...8', 'full_address', 'needs_geo', 'Unnamed: 10'],
      dtype='str')

In [201]:
angelina_data["Zip"] = (
    angelina_data["Zip"]
    .apply(lambda x: str(int(x)) if pd.notna(x) else "")
)

In [202]:
angelina_data.head()

,Venue_Name,Longitude,Latitude,Address,Zip,City,State,...8,full_address,needs_geo,Unnamed: 10
0,1033 Club,-98.480763,29.436116,1033 Avenue B,78215,San Antonio,Texas,NaN,"1033 Avenue B, San Antonio, Texas, 78215",False,NaN
1,151 Saloon,-98.690386,29.464097,10619 Westover Hills Blvd,78215,San Antonio,Texas,NaN,"10619 Westover Hills Blvd, San Antonio, Texas,...",False,NaN
2,1902 Nightclub,-98.478216,29.421548,1174 East Commerce St,78205,San Antonio,Texas,NaN,"1174 East Commerce St, San Antonio, Texas, 78205",True,NaN
3,210 Kapone's,-98.478815,29.42515,1223 East Houston,78205,San Antonio,Texas,NaN,"1223 East Houston, San Antonio, Texas, 78205",True,NaN
4,4th Quarter Sports Bar,-98.569389,29.524564,8779 Wurzbach Rd,78240,San Antonio,Texas,NaN,"8779 Wurzbach Rd, San Antonio, Texas, 78240",False,NaN


I am now ready to compare the student worker's data to Angelina's data, and additionally compare Angelina's data to the venues data in `venues_separated`, since Angelina queried tidyverse whereas I queried Mapbox, so there may be differences that need to be resolved. Once this is done, I will want to match all the coordinate data to the full venues list we now have and re-examine how many venues we have with no data.

### Compare Student and Angelina Coordinates

I want to determine the level of agreement between the student worker's coordinates, derived through manual search, and Angelina's coordinates, derived from an R library query. I generate a script that loads the source data, the student's data, and Angelina's data, and compares them. I set a tolerance of about 22 meters to signal "difference". Only 2 venues have coordinates in both sets within this tolerance. There are seven venues for which only the student found coordinates, and five venues for which only Angelina found coordinates. There are 42 venues for which the coordinates differ by more than 22 meters. 

In [296]:
import pandas as pd
import numpy as np

# ── Load your dataframes ─────────────────────────────────────────────────────
original_df = layer_1_venues  # 974 rows
person1_df  = student_data    # 58 rows, has ID column
person2_df  = angelina_data   # 974 rows, same order as original

# ── Build a comparison frame keyed on original row position ──────────────────
p2_coords = person2_df[["Latitude", "Longitude"]].rename(
    columns={"Latitude": "lat_angelina", "Longitude": "lon_angelina"}
)
p2_coords["_orig_idx"] = p2_coords.index

person1_df["_orig_idx"] = person1_df["ID"].astype(int) - 1
p1_coords = person1_df[["_orig_idx", "LATITUDE", "LONGITUDE"]].rename(
    columns={"LATITUDE": "lat_student", "LONGITUDE": "lon_student"}
)

merged = p1_coords.merge(p2_coords, on="_orig_idx", how="inner")
merged["Venue_Name"] = original_df.loc[merged["_orig_idx"].values, "Venue_Name"].values

# ── Convert to numeric ────────────────────────────────────────────────────────
for col in ["lat_student", "lon_student", "lat_angelina", "lon_angelina"]:
    merged[col] = pd.to_numeric(merged[col], errors="coerce")

# ── Flag differences ──────────────────────────────────────────────────────────
TOLERANCE = 0.0002

merged["lat_diff"]      = (merged["lat_student"] - merged["lat_angelina"]).abs()
merged["lon_diff"]      = (merged["lon_student"] - merged["lon_angelina"]).abs()
merged["both_have_geo"] = merged[["lat_student", "lon_student",
                                   "lat_angelina", "lon_angelina"]].notna().all(axis=1)

# ── Haversine formula ─────────────────────────────────────────────────────────
def haversine_metres(lat1, lon1, lat2, lon2):
    R = 6_371_000
    phi1, phi2 = np.radians(lat1), np.radians(lat2)
    dphi    = np.radians(lat2 - lat1)
    dlambda = np.radians(lon2 - lon1)
    a = np.sin(dphi / 2)**2 + np.cos(phi1) * np.cos(phi2) * np.sin(dlambda / 2)**2
    return R * 2 * np.arctan2(np.sqrt(a), np.sqrt(1 - a))

merged["distance_m"] = np.where(
    merged["both_have_geo"],
    haversine_metres(
        merged["lat_student"], merged["lon_student"],
        merged["lat_angelina"], merged["lon_angelina"]
    ),
    np.nan
)

merged["coords_differ"] = merged["both_have_geo"] & (merged["distance_m"] > 22)  # metres threshold

# ── Subset views ──────────────────────────────────────────────────────────────
conflicts = merged[merged["coords_differ"]]
agree     = merged[merged["both_have_geo"] & ~merged["coords_differ"]]
only_p1   = merged[merged["lat_angelina"].isna() & merged["lat_student"].notna()]
only_p2   = merged[merged["lat_student"].isna() & merged["lat_angelina"].notna()]

# ── Display helper ────────────────────────────────────────────────────────────
def print_comparison(df, cols):
    display = df[cols].copy()
    display["Venue_Name"] = display["Venue_Name"].str[:30]
    print(display.to_string(index=False, float_format=lambda x: f"{x:.2f}"))

# ── Summary ───────────────────────────────────────────────────────────────────
print(f"Venues compared:            {len(merged)}")
print(f"  Both have coordinates:    {merged['both_have_geo'].sum()}")
print(f"  Agree (within tolerance): {len(agree)}")
print(f"  DIFFER:                   {len(conflicts)}")
print(f"  Only Student has coords:  {len(only_p1)}")
print(f"  Only Angelina has coords: {len(only_p2)}")

print("\n── Conflicts ───────────────────────────────────────────────────────────")
print_comparison(conflicts, ["_orig_idx", "Venue_Name", "lat_student", "lon_student",
                              "lat_angelina", "lon_angelina", "distance_m"])

print("\n── Only Student has coordinates ────────────────────────────────────────")
print_comparison(only_p1, ["_orig_idx", "Venue_Name", "lat_student", "lon_student"])

print("\n── Only Angelina has coordinates ───────────────────────────────────────")
print_comparison(only_p2, ["_orig_idx", "Venue_Name", "lat_angelina", "lon_angelina"])

print("\n── Agree (within tolerance) ────────────────────────────────────────────")
print_comparison(agree, ["_orig_idx", "Venue_Name",
                          "lat_student", "lon_student", "lat_angelina", "lon_angelina"])

# ── Export ────────────────────────────────────────────────────────────────────
# conflicts.to_csv("conflicts.csv", index=False)
# merged.to_csv("full_comparison.csv", index=False)

Venues compared:            58
  Both have coordinates:    44
  Agree (within tolerance): 2
  DIFFER:                   42
  Only Student has coords:  7
  Only Angelina has coords: 5

── Conflicts ───────────────────────────────────────────────────────────
 _orig_idx                     Venue_Name  lat_student  lon_student  lat_angelina  lon_angelina  distance_m
       319       Ivory Lounge & Nightclub        29.49       -98.45         29.49        -98.45      152.36
       346                   Just Country        30.15       -94.64         29.52        -98.41   370120.51
       351            Kelly Field Terrace        32.40      -100.87         29.39        -98.49   404246.39
       356                  Keystone Club        30.28       -98.87         29.50        -98.56    91021.93
       365  Krueger's Bar and Party House        30.15       -94.64         29.39        -98.36   368069.09
       372             La Rumba Nite Club        29.71       -96.56         29.43        -98.57

Inspecting the data, I see in some cases differences of hundreds of kilometers between the student's coordinates and Angelina's coordinates. I will add some columns that show how far from the center of San Antonio each set of coordinates is. 

In [297]:
import pandas as pd
import numpy as np

# ── Load your dataframes ─────────────────────────────────────────────────────
original_df = layer_1_venues   # 974 rows
person1_df  = student_data     # 58 rows, has ID column
person2_df  = angelina_data    # 974 rows, same order as original

# ── Build a comparison frame keyed on original row position ──────────────────
p2_coords = person2_df[["Latitude", "Longitude"]].rename(
    columns={"Latitude": "lat_ang", "Longitude": "lon_ang"}
)
p2_coords["_orig_idx"] = p2_coords.index

person1_df["_orig_idx"] = person1_df["ID"].astype(int) - 1
p1_coords = person1_df[["_orig_idx", "LATITUDE", "LONGITUDE"]].rename(
    columns={"LATITUDE": "lat_student", "LONGITUDE": "lon_student"}
)

merged = p1_coords.merge(p2_coords, on="_orig_idx", how="inner")
merged["Venue_Name"] = original_df.loc[merged["_orig_idx"].values, "Venue_Name"].values

# ── Convert to numeric ────────────────────────────────────────────────────────
for col in ["lat_student", "lon_student", "lat_ang", "lon_ang"]:
    merged[col] = pd.to_numeric(merged[col], errors="coerce")

# ── Flag differences ──────────────────────────────────────────────────────────
TOLERANCE = 0.0002

merged["lat_diff"]      = (merged["lat_student"] - merged["lat_ang"]).abs()
merged["lon_diff"]      = (merged["lon_student"] - merged["lon_ang"]).abs()
merged["both_have_geo"] = merged[["lat_student", "lon_student",
                                   "lat_ang", "lon_ang"]].notna().all(axis=1)

# ── Haversine formula ─────────────────────────────────────────────────────────
def haversine_metres(lat1, lon1, lat2, lon2):
    R = 6_371_000
    phi1, phi2 = np.radians(lat1), np.radians(lat2)
    dphi    = np.radians(lat2 - lat1)
    dlambda = np.radians(lon2 - lon1)
    a = np.sin(dphi / 2)**2 + np.cos(phi1) * np.cos(phi2) * np.sin(dlambda / 2)**2
    return R * 2 * np.arctan2(np.sqrt(a), np.sqrt(1 - a))

merged["distance_m"] = np.where(
    merged["both_have_geo"],
    haversine_metres(
        merged["lat_student"], merged["lon_student"],
        merged["lat_ang"], merged["lon_ang"]
    ),
    np.nan
)

merged["coords_differ"] = merged["both_have_geo"] & (merged["distance_m"] > 22)

# ── Distance from central San Antonio ────────────────────────────────────────
SA_LAT = 29.42861139590851
SA_LON = -98.49619652095998

merged["student_sa_dist"] = np.where(
    merged[["lat_student", "lon_student"]].notna().all(axis=1),
    haversine_metres(merged["lat_student"], merged["lon_student"], SA_LAT, SA_LON),
    np.nan
)

merged["ang_sa_dist"] = np.where(
    merged[["lat_ang", "lon_ang"]].notna().all(axis=1),
    haversine_metres(merged["lat_ang"], merged["lon_ang"], SA_LAT, SA_LON),
    np.nan
)

# ── Subset views ──────────────────────────────────────────────────────────────
conflicts = merged[merged["coords_differ"]]
agree     = merged[merged["both_have_geo"] & ~merged["coords_differ"]]
only_p1   = merged[merged["lat_ang"].isna() & merged["lat_student"].notna()]
only_p2   = merged[merged["lat_student"].isna() & merged["lat_ang"].notna()]

# ── Display helper ────────────────────────────────────────────────────────────
def print_comparison(df, cols):
    display = df[cols].copy()
    display["Venue_Name"] = display["Venue_Name"].str[:30]
    print(display.to_string(index=False, float_format=lambda x: f"{x:.2f}"))

# ── Summary ───────────────────────────────────────────────────────────────────
print(f"Venues compared:             {len(merged)}")
print(f"  Both have coordinates:     {merged['both_have_geo'].sum()}")
print(f"  Agree (within tolerance):  {len(agree)}")
print(f"  DIFFER:                    {len(conflicts)}")
print(f"  Only Student has coords:   {len(only_p1)}")
print(f"  Only Angelina has coords:  {len(only_p2)}")

print("\n── Conflicts ───────────────────────────────────────────────────────────")
print_comparison(conflicts, ["_orig_idx", "Venue_Name", "lat_student", "lon_student",
                              "lat_ang", "lon_ang", "distance_m",
                              "student_sa_dist", "ang_sa_dist"])

print("\n── Only Student has coordinates ────────────────────────────────────────")
print_comparison(only_p1, ["_orig_idx", "Venue_Name", "lat_student", "lon_student"])

print("\n── Only Angelina has coordinates ───────────────────────────────────────")
print_comparison(only_p2, ["_orig_idx", "Venue_Name", "lat_ang", "lon_ang"])

print("\n── Agree (within tolerance) ────────────────────────────────────────────")
print_comparison(agree, ["_orig_idx", "Venue_Name",
                          "lat_student", "lon_student", "lat_ang", "lon_ang"])

# ── Export ────────────────────────────────────────────────────────────────────
# conflicts.to_csv("conflicts.csv", index=False)
# merged.to_csv("full_comparison.csv", index=False)

Venues compared:             58
  Both have coordinates:     44
  Agree (within tolerance):  2
  DIFFER:                    42
  Only Student has coords:   7
  Only Angelina has coords:  5

── Conflicts ───────────────────────────────────────────────────────────
 _orig_idx                     Venue_Name  lat_student  lon_student  lat_ang  lon_ang  distance_m  student_sa_dist  ang_sa_dist
       319       Ivory Lounge & Nightclub        29.49       -98.45    29.49   -98.45      152.36          8283.81      8413.75
       346                   Just Country        30.15       -94.64    29.52   -98.41   370120.51        380348.31     13443.12
       351            Kelly Field Terrace        32.40      -100.87    29.39   -98.49   404246.39        400714.71      3994.79
       356                  Keystone Club        30.28       -98.87    29.50   -98.56    91021.93        100871.95     10258.31
       365  Krueger's Bar and Party House        30.15       -94.64    29.39   -98.36   368069.09

Based on this analysis, it would appear that many of the student worker's coordinates are hundreds of kilometers away from San Antonio, whereas Angelina's coordinates are broadly speaking in the San Antonio area. This suggests the student's geolocations are, generally speaking, unreliable and Angelina's tidyverse data should be preferred for the final data set.

In order to record and track these multiple versions, I will write them to an Excel file aimed at internal documentation rather than public sharing. I generate a method to write data frames to a spreadsheet, creating it if it doesn't exist.

In [242]:
import os

def write_df_to_sheet(file_name, sheet_name, df):
    mode = "a" if os.path.exists(file_name) else "w"
    kwargs = {"if_sheet_exists": "replace"} if mode == "a" else {}
    with pd.ExcelWriter(file_name, engine="openpyxl", mode=mode, **kwargs) as writer:
        df.to_excel(writer, sheet_name=sheet_name, index=False)

In [275]:
# write_df_to_sheet('venues_work.xlsx', 'layer_1_venues_original', layer_1_venues)

In [276]:
# mapbox_data = pd.read_excel('spreadsheets/venues_separated.xlsx', sheet_name='All Venues')
# write_df_to_sheet('venues_work.xlsx', 'layer_1_venues_mapbox_data', mapbox_data)

In [298]:
# write_df_to_sheet('venues_work.xlsx', 'student_coords', student_data)

In [278]:
# write_df_to_sheet('venues_work.xlsx', 'angelina_tidyverse_coords', angelina_data)

In [299]:
# write_df_to_sheet('venues_work.xlsx', 'student_angelina_comparison', merged)

In [300]:
# write_df_to_sheet('venues_work.xlsx', 'student_angelina_differences', conflicts)

### Compare Tidyverse and Mapbox Data

One more bit of analysis may prove of interest. Although the Mapbox API obtained coordinates for all but 71 locations in March 2024, since the Layer 1 Venues spreadsheet was not updated, Angelina Cortes was given a Layer 1 Venues sheet missing over 600 locations, and ended up duplicating the API call, but using a different library, Tidyverse. It may be instructive to compare the two sets of coordinates.

In [281]:
import pandas as pd
import numpy as np
import os

# ── Load your dataframes ─────────────────────────────────────────────────────
original_df  = layer_1_venues  # 974 rows
mapbox_df    = mapbox_data     # 974 rows, has id column (id = orig_idx + 1)
angelina_df  = angelina_data   # 974 rows, same order as original

# ── Haversine formula ─────────────────────────────────────────────────────────
def haversine_metres(lat1, lon1, lat2, lon2):
    R = 6_371_000
    phi1, phi2 = np.radians(lat1), np.radians(lat2)
    dphi    = np.radians(lat2 - lat1)
    dlambda = np.radians(lon2 - lon1)
    a = np.sin(dphi / 2)**2 + np.cos(phi1) * np.cos(phi2) * np.sin(dlambda / 2)**2
    return R * 2 * np.arctan2(np.sqrt(a), np.sqrt(1 - a))

# ── San Antonio centre ────────────────────────────────────────────────────────
SA_LAT = 29.42861139590851
SA_LON = -98.49619652095998

# ── Build comparison frame keyed on original row position ────────────────────
# Mapbox: id = orig_idx + 1
mapbox_df["_orig_idx"] = mapbox_df["id"].astype(int) - 1
mb_coords = mapbox_df[["_orig_idx", "latitude", "longitude"]].rename(
    columns={"latitude": "mapbox_lat", "longitude": "mapbox_long"}
)

# Angelina: already in original order
ang_coords = angelina_df[["Latitude", "Longitude"]].rename(
    columns={"Latitude": "tidyverse_lat", "Longitude": "tidyverse_long"}
)
ang_coords["_orig_idx"] = ang_coords.index

# ── Join on position ──────────────────────────────────────────────────────────
merged = mb_coords.merge(ang_coords, on="_orig_idx", how="inner")
merged["Venue_Name"] = original_df.loc[merged["_orig_idx"].values, "Venue_Name"].values

# ── Convert to numeric ────────────────────────────────────────────────────────
for col in ["mapbox_lat", "mapbox_long", "tidyverse_lat", "tidyverse_long"]:
    merged[col] = pd.to_numeric(merged[col], errors="coerce")

# ── Lat/lon diffs ─────────────────────────────────────────────────────────────
TOLERANCE = 0.0002

merged["lat_diff"] = (merged["mapbox_lat"] - merged["tidyverse_lat"]).abs()
merged["lon_diff"] = (merged["mapbox_long"] - merged["tidyverse_long"]).abs()
merged["both_have_geo"] = merged[["mapbox_lat", "mapbox_long",
                                   "tidyverse_lat", "tidyverse_long"]].notna().all(axis=1)

# ── Distance between the two results ─────────────────────────────────────────
merged["distance"] = np.where(
    merged["both_have_geo"],
    haversine_metres(
        merged["mapbox_lat"], merged["mapbox_long"],
        merged["tidyverse_lat"], merged["tidyverse_long"]
    ),
    np.nan
)

merged["coords_differ"] = merged["both_have_geo"] & (merged["distance"] > 22)  # metres threshold

# ── Distance from central San Antonio ────────────────────────────────────────
merged["mapbox_dist"] = np.where(
    merged[["mapbox_lat", "mapbox_long"]].notna().all(axis=1),
    haversine_metres(merged["mapbox_lat"], merged["mapbox_long"], SA_LAT, SA_LON),
    np.nan
)

merged["tidyverse_dist"] = np.where(
    merged[["tidyverse_lat", "tidyverse_long"]].notna().all(axis=1),
    haversine_metres(merged["tidyverse_lat"], merged["tidyverse_long"], SA_LAT, SA_LON),
    np.nan
)

# ── Subset views ──────────────────────────────────────────────────────────────
conflicts    = merged[merged["coords_differ"]]
agree        = merged[merged["both_have_geo"] & ~merged["coords_differ"]]
only_mapbox  = merged[merged["tidyverse_lat"].isna() & merged["mapbox_lat"].notna()]
only_tidy    = merged[merged["mapbox_lat"].isna() & merged["tidyverse_lat"].notna()]

# ── Summary ───────────────────────────────────────────────────────────────────
print(f"Venues compared:              {len(merged)}")
print(f"  Both have coordinates:      {merged['both_have_geo'].sum()}")
print(f"  Agree (within tolerance):   {len(agree)}")
print(f"  DIFFER:                     {len(conflicts)}")
print(f"  Only Mapbox has coords:     {len(only_mapbox)}")
print(f"  Only Tidyverse has coords:  {len(only_tidy)}")

def print_comparison(df, cols):
    display = df[cols].copy()
    display["Venue_Name"] = display["Venue_Name"].str[:30]
    print(display.to_string(index=False, float_format=lambda x: f"{x:.2f}"))


print("\n── Conflicts ───────────────────────────────────────────────────────────")
print_comparison(conflicts, ["_orig_idx", "Venue_Name",
                             "mapbox_lat", "mapbox_long",
                             "tidyverse_lat", "tidyverse_long",
                             "distance", "mapbox_dist", "tidyverse_dist"])

print("\n── Only Mapbox has coordinates ─────────────────────────────────────────")
print(only_mapbox[["_orig_idx", "Venue_Name", "mapbox_lat", "mapbox_long"]].to_string(index=False))

print("\n── Only Tidyverse has coordinates ──────────────────────────────────────")
print(only_tidy[["_orig_idx", "Venue_Name", "tidyverse_lat", "tidyverse_long"]].to_string(index=False))

print("\n── Agree (within tolerance) ────────────────────────────────────────────")
print(agree[["_orig_idx", "Venue_Name",
             "mapbox_lat", "mapbox_long",
             "tidyverse_lat", "tidyverse_long"]].to_string(index=False))

# ── Export ────────────────────────────────────────────────────────────────────
# conflicts.to_csv("conflicts_mapbox_vs_tidyverse.csv", index=False)
# merged.to_csv("full_comparison_mapbox_vs_tidyverse.csv", index=False)

Venues compared:              974
  Both have coordinates:      892
  Agree (within tolerance):   348
  DIFFER:                     544
  Only Mapbox has coords:     10
  Only Tidyverse has coords:  70

── Conflicts ───────────────────────────────────────────────────────────
 _orig_idx                     Venue_Name  mapbox_lat  mapbox_long  tidyverse_lat  tidyverse_long    distance  mapbox_dist  tidyverse_dist
         2                 1902 Nightclub       29.42       -98.48          29.42          -98.48      104.64      1952.86         1910.38
         3                   210 Kapone's       29.43       -98.48          29.43          -98.48       36.61      1761.28         1726.79
         6                    Abracadabra       29.42       -98.50          29.52          -98.50    10596.48       457.61        10145.12
         7                 Acapulco Sam's       29.43       -98.49          29.43          -98.49       34.24       749.33          715.44
         8                   

In [274]:
# write_df_to_sheet('venues_work.xlsx', 'mapbox_tidyverse_comparison', merged)
# write_df_to_sheet('venues_work.xlsx', 'mapbox_tidyverse_differences', conflicts)

You can see that in cases where there's large differences between the two data sets, it would appear that Tidyverse is more reliable. For example, Alexander's Bar and Café appears to be almost 4000 km from San Antonio according to Mapbox; Kelly Field Terrace is 400 km away, and Keystone Club is 100 km away. In all cases, Tidyverse offers the more likely coordinates. In addition, Tidyverse was more complete: it has coordinates for seventy locations that Mapbox could not locate. Because of its higher degree of accuracy and completeness, Tidyverse results are therefore to be preferred over Mapbox.

It is also of interest that there are 348 venues for which the two APIs generated results within 22 meters of each other.

### Generate comprehensive geolocation table

I've analyzed all three sets of geolocation data, and evaluated their reliability and completeness. The next step is to combine the geolocation data with the comprehensive venues list. The resulting table should have the known data on address and geolocation along with the venue name and identifier.